tests/unit/test_charm.py

```diff
@@ -181,6 +231,89 @@ def test_on_config_changed_iscsi(harness, mocker, iscsi_config):
+        (INITIATOR_CONTENT_TEMPLATE.format("InitiatorName=iqn.2020-07.canonical.com:lun1"), False),
+    ],
+    ids=["no-file", "empty-file", "different-iqn", "same-iqn"],
+)
+def test_iscsi_with_initiator_dictionary_hostname(
+    mocker, harness, iscsi_config, initiator_content, render_expected
+):
+    """Test iscsi configuration when hostname is present in initiator dicitionary."""
+    mocker.patch("charm.utils.is_container", return_value=False)
+    mocker.patch("charm.subprocess.getoutput", return_value="iqn.2020-07.canonical.com:lun1")
+    mocker.patch("charm.socket.getfqdn", return_value="testhost.testdomain")
+    mocker.patch("charm.subprocess.check_call")
+    mocker.patch("charm.subprocess.check_output")
+    mocker.patch("charm.StorageConnectorCharm._configure_deferred_restarts")
+    mocker.patch("charm.StorageConnectorCharm._defer_service_restart")
+    mocker.patch("charm.Path.chmod")
+    mock_render = mocker.patch("charm.StorageConnectorCharm._render_iscsi_initiator")
+
+    if initiator_content is not None:
+        harness.charm.ISCSI_INITIATOR_NAME.write_text(initiator_content)
+
+    harness.update_config(iscsi_config)
+
+    if render_expected:
+        mock_render.assert_called_once()
+    else:
+        mock_render.assert_not_called()
+
+    assert isinstance(harness.charm.unit.status, ActiveStatus)
+
+
+@pytest.mark.parametrize(
+    "initiator_content, render_expected",
+    [
+        (None, True),  # file doesn't exist
+        ("", True),  # file exists but no content
+        (INITIATOR_CONTENT_TEMPLATE.format("InitiatorName=test-iqn"), False),
+    ],
+    ids=["no-file", "empty-file", "content-present"],
+)
+def test_iscsi_without_initiator_dictionary_hostname(
+    mocker, harness, iscsi_config, initiator_content, render_expected
+):
+    """Test iscsi config when hostname is absent from initiator-dictionary."""
+    mocker.patch("charm.utils.is_container", return_value=False)
+    mock_getoutput = mocker.patch(
+        "charm.subprocess.getoutput", return_value="iqn.2020-07.canonical.com:lun1"
+    )
+    mocker.patch("charm.socket.getfqdn", return_value="testhost.testdomain")
+    mocker.patch("charm.subprocess.check_call")
+    mocker.patch("charm.subprocess.check_output")
+    mocker.patch("charm.StorageConnectorCharm._configure_deferred_restarts")
+    mocker.patch("charm.StorageConnectorCharm._defer_service_restart")
+    mocker.patch("charm.Path.chmod")
+    mock_getoutput = mocker.patch(
+        "charm.subprocess.getoutput", return_value="iqn.2020-07.canonical.com:lun1"
+    )
+    mock_render = mocker.patch("charm.StorageConnectorCharm._render_iscsi_initiator")
+
+    if initiator_content is not None:
+        harness.charm.ISCSI_INITIATOR_NAME.write_text(initiator_content)
+
+    iscsi_config["initiator-dictionary"] = "{}"
+    harness.update_config(iscsi_config)
+
+    if render_expected:
+        mock_render.assert_called_once()
```

Comment:
can we use assert_called_once_with here to make sure it will write the correct initiator_name into file